ddtrace/profiling/collector/_memalloc_heap.c

```diff
@@ -56,12 +65,72 @@ memalloc_heap_untrack(void* ptr)
         if (ptr == (*tb)->ptr) {
             /* Free the traceback */
             traceback_free(*tb);
-            traceback_array_remove(&global_heap_tracker.allocs, tb);
+            traceback_array_remove(&heap_tracker->allocs, tb);
             break;
         }
     }
 }
 
+static void
+heap_tracker_thaw(heap_tracker_t* heap_tracker)
+{
+    /* Add the frozen allocs at the end */
+    traceback_array_splice(&heap_tracker->allocs,
+                           heap_tracker->allocs.count,
+                           0,
+                           heap_tracker->freezer.allocs.tab,
+                           heap_tracker->freezer.allocs.count);
+
+    /* Handle the frees: we need to handle the frees after we merge the allocs
+       array together to be sure that there's no free in the freezer matching
+       an alloc that is also in the freezer; heap_tracker_untrack_thawed does
+       not care about the freezer, by definition. */
+    for (TRACEBACK_ARRAY_COUNT_TYPE i = 0; i < heap_tracker->freezer.frees.count; i++)
+        heap_tracker_untrack_thawed(heap_tracker, heap_tracker->freezer.frees.tab[i]);
+
+    /* Reset the count to zero so we can reused the array and overwrite previous values */
+    heap_tracker->freezer.allocs.count = 0;
+    heap_tracker->freezer.frees.count = 0;
+
+    heap_tracker->frozen = false;
+}
+
+/* Public API */
+
+void
+memalloc_heap_tracker_init(void)
+{
+    heap_tracker_init(&global_heap_tracker);
+}
+
+void
+memalloc_heap_tracker_deinit(void)
+{
+    heap_tracker_wipe(&global_heap_tracker);
+}
+
+void
+memalloc_heap_tracker_freeze()
+{
+    heap_tracker_freeze(&global_heap_tracker);
+}
+
+void
+memalloc_heap_tracker_thaw()
+{
+    heap_tracker_thaw(&global_heap_tracker);
+}
+
+void
+memalloc_heap_untrack(void* ptr)
+{
+    if (global_heap_tracker.frozen)
+        /* FIXME check for max capacity of frees */
```

Comment:
Worth doing still to be safe? It is unlikely that there will be very many frees while frozen so probably not a big concern, but then again, shouldn't be too hard to check either 🤷‍♂️